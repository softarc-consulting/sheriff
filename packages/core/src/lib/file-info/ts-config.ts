import { CompilerOptions } from 'typescript';

export interface TsConfig {
  compilerOptions: CompilerOptions;
  extends: string;
}
