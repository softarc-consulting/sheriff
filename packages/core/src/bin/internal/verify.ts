import * as path from 'path';
import * as process from 'process';
import { init } from '../../lib/main/init';
import { FsPath, toFsPath } from '../../lib/file-info/fs-path';
import { traverseFileInfo } from '../../lib/modules/traverse-file-info';
import { hasDeepImport } from '../../lib/eslint/deep-import';
import { checkForDeepImports } from '../../lib/checks/check-for-deep-imports';
import {
  checkForDependencyRuleViolation,
  DependencyRuleViolation,
} from '../../lib/checks/check-for-dependency-rule-violation';

export function verify(args: string[]) {
  const [main] = args;
  const mainPath = path.join(path.resolve(process.cwd(), main));
  const projectConfig = init(toFsPath(mainPath));

  let deepImportsCount = 0
  let dependencyRulesCount = 0
  let filesCount = 0

  let hasError = false;

  for (const { fileInfo } of traverseFileInfo(projectConfig.fileInfo)) {
    const deepImports = checkForDeepImports(fileInfo.path, projectConfig);
    const dependencyRuleViolations = checkForDependencyRuleViolation(
      fileInfo.path,
      projectConfig,
    );

    if (deepImports.length > 0 || dependencyRuleViolations.length > 0) {
      hasError = true;
      filesCount++

      console.log(fileInfo.path);
      console.log('-'.repeat(fileInfo.path.length));

      if (deepImports.length > 0) {
        console.log('Deep Imports');
        deepImportsCount += deepImports.length
        deepImports.forEach((deepImport) => {
          console.log('  ' + deepImport);
        });
      }

      if (dependencyRuleViolations.length > 0) {
        console.log('Dependency Rule Violations');
        dependencyRulesCount += dependencyRuleViolations.length
        dependencyRuleViolations.forEach((dependencyRuleViolation) => {
          console.log(
            '  from tags:' +
              dependencyRuleViolation.fromTags.join(', ') +
              ', to:' +
              dependencyRuleViolation.toTag,
          );
        });
      }
    }
  }

  if (hasError) {
    console.error("Issues found in");
    console.error(`  Files: ${filesCount}`);
    console.error(`  Deep Imports: ${deepImportsCount}`);
    console.error(`  Dependency Rule Violations: ${dependencyRulesCount}`);
    process.exit(1); // failure
  }
  else {
    console.log('')
    console.log('\u001b[32mNo issues found. The project is clean.')
  }
}
