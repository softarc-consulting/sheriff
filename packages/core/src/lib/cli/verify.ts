import * as path from 'path';
import { toFsPath } from '../file-info/fs-path';
import { init } from '../main/init';
import { checkForDeepImports } from '../checks/check-for-deep-imports';
import { traverseFileInfo } from '../modules/traverse-file-info';
import { checkForDependencyRuleViolation } from '../checks/check-for-dependency-rule-violation';
import { Cli } from './util';
import { UserError } from '../error/user-error';
import getFs from '../fs/getFs';

type ValidationsMap = Record<
  string,
  { deepImports: string[]; dependencyRules: string[] }
>;

export function verify(currentDir: string, cli: Cli, args: string[]) {
  let deepImportsCount = 0;
  let dependencyRulesCount = 0;
  let filesCount = 0;
  let hasError = false;
  const validationsMap: ValidationsMap = {};
  const fs = getFs();

  try {
    const [main] = args;
    const mainPath = path.join(path.resolve(currentDir, main));
    const projectConfig = init(toFsPath(mainPath));

    for (const { fileInfo } of traverseFileInfo(projectConfig.fileInfo)) {
      const deepImports = checkForDeepImports(fileInfo.path, projectConfig);
      const dependencyRuleViolations = checkForDependencyRuleViolation(
        fileInfo.path,
        projectConfig,
      );

      if (deepImports.length > 0 || dependencyRuleViolations.length > 0) {
        hasError = true;
        filesCount++;
        deepImportsCount += deepImports.length;
        dependencyRulesCount += dependencyRuleViolations.length;

        const dependencyRules = dependencyRuleViolations.map(
          (violation) =>
            `from tags ${violation.fromTags.join(',')} to ${violation.toTag}`,
        );

        validationsMap[fs.relativeTo(currentDir, fileInfo.path)] = {
          deepImports,
          dependencyRules,
        };
      }
    }

    cli.log('');
    cli.log('\u001b[1mVerification Report\u001b[0m');
    if (hasError) {
      cli.log('');
      cli.log('Issues found:');
      cli.log(`  Total Invalid Files: ${filesCount}`);
      cli.log(`  Total Deep Imports: ${deepImportsCount}`);
      cli.log(`  Total Dependency Rule Violations: ${dependencyRulesCount}`);
      cli.log('----------------------------------');
      cli.log('');
    } else {
      cli.log('');
      cli.log('\u001b[32mNo issues found. Well done!');
      cli.endProcessOk();
    }

    for (const [file, { deepImports, dependencyRules }] of Object.entries(
      validationsMap,
    )) {
      cli.log('|-- ' + file);
      if (deepImports.length > 0) {
        cli.log('|   |-- Deep Imports');
        deepImports.forEach((deepImport) => {
          cli.log('|   |   |-- ' + deepImport);
        });
      }

      if (dependencyRules.length > 0) {
        cli.log('|   |-- Dependency Rule Violations');
        dependencyRules.forEach((dependencyRule) => {
          cli.log('|   |   |-- ' + dependencyRule);
        });
      }
    }

    cli.endProcessError();
  } catch (error) {
    if (error instanceof UserError) {
      cli.logError(error.message);
    } else {
      cli.logError(String(error));
    }
    cli.endProcessError();
  }
}
