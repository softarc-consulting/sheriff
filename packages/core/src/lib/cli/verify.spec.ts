import { beforeEach, describe, expect, vitest } from 'vitest';
import { verify } from './verify';
import { createProject } from '../test/project-creator';
import { sheriffConfig } from '../test/project-configurator';
import { tsConfig } from '../test/fixtures/ts-config';

const setup = () => {
  const cli = {
    endProcessOk: vitest.fn(),
    endProcessError: vitest.fn(),
    log: vitest.fn<void[string]>(),
    logError: vitest.fn<void, [string]>(),
  };

  const allLogs = () =>
    cli.log.mock.calls.map(([message]) => message).join('\n');
  const allErrorLogs = () =>
    cli.logError.mock.calls.map(([message]) => message).join('\n');

  return { cli, allLogs, allErrorLogs };
};

describe('verify', (it) => {
  beforeEach(() => {
    vitest.resetAllMocks();
  });

  it('should find no errors', () => {
    const { cli, allLogs, allErrorLogs } = setup();

    createProject({
      'tsconfig.json': tsConfig(),
      src: {
        'main.ts': [''],
      },
    });

    verify('/project', cli, ['src/main.ts']);

    expect(allErrorLogs()).toMatchSnapshot('error');
    expect(allLogs()).toMatchSnapshot('log');
  });

  it('should show user errors', () => {
    const { cli, allLogs, allErrorLogs } = setup();

    createProject({
      'tsconfig.json': tsConfig(),
      'sheriff.config.ts': sheriffConfig({
        tagging: {
          'src/holidays': 'domain:<domain>',
        },
        depRules: {},
      }),
      src: {
        'main.ts': ['./holidays'],
        holidays: {
          'index.ts': [],
        },
      },
    });

    verify('/project', cli, ['src/main.ts']);

    expect(allLogs()).toMatchSnapshot('log');
    expect(allErrorLogs()).toMatchSnapshot('error');
  });

  it('should show internal errors', () => {
    expect(true).toBe(true);
  });

  it('should find errors', () => {
    const { cli, allLogs, allErrorLogs } = setup();

    createProject({
      'tsconfig.json': tsConfig(),
      'sheriff.config.ts': sheriffConfig({
        tagging: {
          'src/customers': ['customers'],
          'src/holidays': ['holidays'],
        },
        depRules: {
          root: ['customers', 'holidays'],
          customers: [],
          holidays: [],
        },
      }),
      src: {
        'main.ts': ['./holidays', './customers', './customers/data'],
        holidays: {
          'index.ts': ['./holidays.component'],
          'holidays.component.ts': ['../customers'],
        },
        customers: { 'index.ts': [], 'data.ts': [] },
      },
    });

    verify('/project', cli, ['src/main.ts']);

    expect(allErrorLogs()).toMatchSnapshot('error.log');
    expect(allLogs()).toMatchSnapshot('logs.log');
  });
});
