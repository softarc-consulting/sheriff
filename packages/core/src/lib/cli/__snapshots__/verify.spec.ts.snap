// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`verify > should find errors > error.log 1`] = `""`;

exports[`verify > should find errors > logs.log 1`] = `
"
[1mVerification Report[0m

Issues found:
  Total Invalid Files: 2
  Total Deep Imports: 1
  Total Dependency Rule Violations: 1
----------------------------------

|-- src/main.ts
|   |-- Deep Imports
|   |   |-- ./customers/data
|-- src/holidays/holidays.component.ts
|   |-- Dependency Rule Violations
|   |   |-- from tags holidays to customers"
`;

exports[`verify > should find no errors > error 1`] = `""`;

exports[`verify > should find no errors > log 1`] = `
"
[1mVerification Report[0m

[32mNo issues found. Well done!"
`;

exports[`verify > should show user errors > error 1`] = `"cannot find a placeholder for \\"domain\\" in tag configuration. Module: /project/src/holidays"`;

exports[`verify > should show user errors > log 1`] = `""`;
